src/keyboard-lomiri.c

```diff
@@ -206,67 +263,97 @@ void keyboard_GetLayout(Keyboard *pKeyboard, gint nLayout, gchar **pLanguage, gc
+            g_settings_set_value (pKeyboard->pPrivate->pSettings, "sources", pSettingsSources);
+
+            GDBusConnection *pConnection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, NULL);
+            gint nUid = geteuid ();
+            gchar *sPath = g_strdup_printf ("/org/freedesktop/Accounts/User%i", nUid);
+            GDBusProxy *pProxy = g_dbus_proxy_new_sync (pConnection, G_DBUS_PROXY_FLAGS_NONE, NULL, "org.freedesktop.Accounts", sPath, "org.freedesktop.DBus.Properties", NULL, NULL);
+            g_free (sPath);
+            GVariant *pAccountsSources = g_variant_builder_end (&cAccountsBuilder);
+
+            #ifdef ENABLE_UBUNTU_ACCOUNTSSERVICE
+                gchar *sInterface = "org.freedesktop.Accounts.User";
+            #else
+                gchar *sInterface = "com.lomiri.shell.AccountsService";
+            #endif
+
+            GVariant *pRet = g_dbus_proxy_call_sync (pProxy, "Set", g_variant_new ("(ssv)", sInterface, "InputSources", pAccountsSources), G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL);
+            g_variant_unref (pRet);
+            g_object_unref (pConnection);
         }
+        else
+        {
+            gchar *sId = g_slist_nth_data (pKeyboard->pPrivate->lLayoutRecSoftware, nLayout);
+            guint nId = 0;
+            gchar *sLayout = NULL;
 
-        GVariant *pSettingsSources = g_variant_builder_end (&cSettingsBuilder);
-        g_settings_set_value (pKeyboard->pPrivate->pSettings, "sources", pSettingsSources);
+            while (LOMIRI_TO_ISO[nId][0] != NULL)
+            {
+                gboolean bEqual = g_str_equal (LOMIRI_TO_ISO[nId][1], sId);
 
-        GDBusConnection *pConnection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, NULL);
-        gint nUid = geteuid ();
-        gchar *sPath = g_strdup_printf ("/org/freedesktop/Accounts/User%i", nUid);
-        GDBusProxy *pProxy = g_dbus_proxy_new_sync (pConnection, G_DBUS_PROXY_FLAGS_NONE, NULL, "org.freedesktop.Accounts", sPath, "org.freedesktop.DBus.Properties", NULL, NULL);
-        g_free (sPath);
-        GVariant *pAccountsSources = g_variant_builder_end (&cAccountsBuilder);
+                if (bEqual)
+                {
+                    sLayout = LOMIRI_TO_ISO[nId][0];
 
-        #ifdef ENABLE_UBUNTU_ACCOUNTSSERVICE
-            gchar *sInterface = "org.freedesktop.Accounts.User";
-        #else
-            gchar *sInterface = "com.lomiri.shell.AccountsService";
-        #endif
+                    break;
+                }
 
-        GVariant *pRet = g_dbus_proxy_call_sync (pProxy, "Set", g_variant_new ("(ssv)", sInterface, "InputSources", pAccountsSources), G_DBUS_CALL_FLAGS_NONE, -1, NULL, NULL);
-        g_variant_unref (pRet);
-        g_object_unref (pConnection);
+                nId++;
+            }
+
+            if (!sLayout)
+            {
+                sLayout = sId;
+            }
+
+            g_settings_set_string (pKeyboard->pPrivate->pMaliitSettings, "active-language", sLayout);
+        }
     }
```

Comment:
Can this be done more patch review friendly? E.g. Introduce two methods/functions: keyboard_SetLayout_OSK and keyboard_SetLayout_HwKeybd and do the ``if (bOSK)`` in a wrapper function keyboard_SetLayout. That would save us the white-space / indentation changes. Sorry for the hassle...